### PR TITLE
change account information feature

### PR DESCRIPTION
**Pull request by waiterorg**
Screenshot
ScreenshotEdit

### CODE REVIEW

**Comment by DutcherNL** [comment]
Having fun with out codebase I see :D

May I ask how you found us and what your interest is in our code?

**Comment by waiterorg** [comment]
I was looking for an interesting project in github
I saw your project and checked it, sound interesting knights board game and roleplaying to me
I wanted to contribute on your project :)

**Comment by DutcherNL** [comment]
Yeah, we are a student association in Eindhoven, the Netherlands, devoted to board- and roleplaying games. Squire is our internal platform where we track our membership as well as our board- and roleplaying collection. It's a hodge podge of features that we deemed useful right now :)
While we do appreciate your contribution, we'd like to wait a bit till we are done and implemented with #197. It combines the several personal pages into one navigatable layout and also results in some code being shifted around, including the bit where you are implementing this. I hope you can understand that.

**Comment by waiterorg** [comment]
That's ok, thank you so much for your explain
I wish best for your team and your platform :)

File: core/views.py [hunk 1]
```diff
@@ -42,6 +46,13 @@ def viewNewsletters(request):
         'NEWSLETTER_ARCHIVE_URL': global_preferences['newsletter__share_link'],
     })
 
+class AccountMixin(LoginRequiredMixin):
+    """
+        Sets the view's object to the user that makes
+        the request.
+    """
+    def get_object(self, queryset=None):
+        return self.request.user
```
Comment (EricTRL) [inline core/views.py:55]: I'd argue that you can move `get_object` straight to `AccountChangeView` as it's not used anywhere else. Furthermore, `AccountChangeview` can then directly inherit `LoginRequiredMixin`

File: core/views.py [hunk 2]
```diff
@@ -74,6 +85,22 @@ def get_context_data(self, *args, **kwargs):
         context[self.tab_name] = True
         return context
 
+
+class AccountChangeView(AccountMixin, AccountTabsMixin, PermissionRequiredMixin, UpdateView):
+    template_name = 'core/user_accounts/account_edit.html'
+    form_class = AccountForm
+    success_url = reverse_lazy('core:user_accounts/account')
+    permission_required = ('core.can_view_account_information_self', 'core.can_change_account_information_self')
```
Comment (EricTRL) [inline core/views.py:93]: These permissions don't actually exist. I suspect that you took the membership change view as a reference point for your changes, which do require specific permissions which do exist. Note how we create those in the model `membership_file.models.Member` using that model's [Meta option for `permissions`](https://docs.djangoproject.com/en/2.2/ref/models/options/#permissions)

Either way, I don't think we would want to limit account editing through permissions, so I'd suggest to remove `PermissionRequiredMixin` and its corresponding required permissions:
```suggestion

```

File: core/views.py [hunk 3]
```diff
@@ -74,6 +85,22 @@ def get_context_data(self, *args, **kwargs):
...
+    raise_exception = True
+    tab_name = 'tab_account'
+
+
+    def form_valid(self, form):
+        message = _("Your account information has been saved successfully! Notice the name disappears when the user is a membership ")
```
Comment (EricTRL) [inline core/views.py:99]: The name-hiding for members is basically an ugly hack on our side; there's no need to mention that to our members
```suggestion
        message = _("Your account information has been saved successfully!")
```

File: core/forms.py [hunk 4]
```diff
@@ -143,3 +143,15 @@ class Meta:
         fields = "__all__"
 
     updating_user_field_name = "uploader"
+
+class AccountForm(ModelForm):
+    class Meta:
+        model = User
+        fields = ['username', 'first_name', 'email']
+
```
Comment (EricTRL) [inline core/forms.py:151]: It's possible to replace the 'first_name' field label by custom text in a [modelform's meta options](https://docs.djangoproject.com/en/3.2/topics/forms/modelforms/#overriding-the-default-fields). For instance, we can change it to match the field name to how it shows up when a user registers:
```suggestion
    labels = {
        'first_name': 'Real Name',
    }
        
```

**Review by EricTRL — CHANGES_REQUESTED** [review]
As @DutcherNL already mentioned, thank you for your contribution. We appreciate it! Now that #197 has been merged we're ready to tackle this feature! Quite some code has moved around, so do ask for help if you're having trouble with finding the relevant files and folders. 

Note that the core of your PR (e.g., the view, form, or templates you created) is still standing strong, and to get this thing running on master will _mostly_ involve moving code around to different files. Again, if you're having trouble with that, do ask for help!

I've also added some specific feedback, which are mostly minor changes.

**Comment by waiterorg** [comment]
Sorry for taking so long to do this actions, thank you so much for your feedback, I did changes which you requested.

**Comment by DutcherNL** [comment]
Not a problem. Just take your time, we do too :)
I do notice that there are still some conflicts that need to be resolved, can you resolve those so we can look at them this weekend?

**Comment by waiterorg** [comment]
what conflicts?
if you know can you tell me please ?

**Comment by DutcherNL** [comment]
At the bottom of this page, github says taht the branch has conflicts with the master that must be resolved. It applies to:
core/templates/core/user_accounts/account_info.html
core/urls.py
core/views.py

File: core/urls.py [hunk 5]
```diff
@@ -57,6 +57,7 @@
             extra_context={},
         ),
         name='user_accounts/password_reset/success'),
+<<<<<<< HEAD
```
Comment (EricTRL) [inline core/urls.py:60]: The account and password-change urls have been moved to `user_interaction/account_pages/config.py`. The account-change url can be moved there as well.

File: core/views.py [hunk 6]
```diff
@@ -6,15 +6,23 @@
 from django.contrib.auth.mixins import LoginRequiredMixin, PermissionRequiredMixin
 from django.contrib.contenttypes.models import ContentType
 from django.core.exceptions import ObjectDoesNotExist
+<<<<<<< HEAD
```
Comment (EricTRL) [inline core/views.py:9]: To resolve these import conflicts, I'd suggest to keep the imports from both your branch and those from master.

File: core/views.py [hunk 7]
```diff
@@ -47,6 +55,7 @@ def viewNewsletters(request):
     })
 
 
+<<<<<<< HEAD
```
Comment (EricTRL) [inline core/views.py:58]: `AccountTabsMixin` actually no longer exists with the new setup. You can move your `AccountChangeView` to `user_interaction/account_pages/views.py` and have your `AccountChangeView` inherit from `AccountViewMixin` instead as well (see how this is done for `SiteAccountView` for instance)

**Review by EricTRL — CHANGES_REQUESTED** [review]
It seems like some things went wrong during the merge as there are some merge conflicts left. I left some review suggestions on how to solve those. :)

I've also enabled the test runner for this PR.
And as Dutcher already mentioned: Don't worry about taking your time. We're doing this as a side-project and there's no rush involved whatsoever :)

File: core/views.py [hunk 8]
```diff
@@ -86,17 +79,18 @@ def get_context_data(self, *args, **kwargs):
...
-    permission_required = ('core.can_view_account_information_self', 'core.can_change_account_information_self')
     raise_exception = True
     tab_name = 'tab_account'
 
+    def get_object(self, queryset=None):
+        return self.request.user
 
     def form_valid(self, form):
-        message = _("Your account information has been saved successfully! Notice the name disappears when the user is a membership ")
+        message = message = _("Your account information has been saved successfully!")
```
Comment (EricTRL) [inline core/views.py:102]: This change also seems like a small error (note the extra `message =`

**Comment by waiterorg** [comment]
Sorry about pervious commit .
I change view , form, template but i  could not figure out where i can implement url .

File: core/templates/core/user_accounts/account_edit.html [hunk 9]
```diff
@@ -0,0 +1,36 @@
+{% extends 'core/user_accounts/account_base.html' %}
```
Comment (EricTRL) [inline core/templates/core/user_accounts/account_edit.html:1]: This file is identical to the other `account_edit.html` file created in this PR. Feel free to remove this one as it's not actually used.

File: user_interaction/templates/user_interaction/account_pages/account_edit.html [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+              <li>{{ error }}</li>
+            {% endfor %}
+          </ul>
+        {% endif %}
+
+            {% generic_field form.username -1 %}
+
+            {% generic_field form.first_name -1 %}
+
+            {% generic_field form.email -1 %}>
```
Comment (EricTRL) [inline user_interaction/templates/user_interaction/account_pages/account_edit.html:24]: There's an additional `>` at the end
```suggestion
            {% generic_field form.email -1 %}
```

File: user_interaction/templates/user_interaction/account_pages/account_edit.html [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+            <p><i> You can change your password <a href="{% url 'core:user_accounts/password_change' %}">here</a></i></p>
```
Comment (EricTRL) [inline user_interaction/templates/user_interaction/account_pages/account_edit.html:26]: The password-change url was moved
```suggestion
            <p><i> You can change your password <a href="{% url 'account:password_change' %}">here</a></i></p>
```

File: user_interaction/templates/user_interaction/account_pages/account_edit.html [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        <input type="submit" value="Confirm Changes" class="greenify btn btn-dark">
+        <a
+        href="{% url 'core:user_accounts/account' %}"
```
Comment (EricTRL) [inline user_interaction/templates/user_interaction/account_pages/account_edit.html:29]: Same as above (and similar to the change you already made in `AccountChangeView.success_url`)
```suggestion
        href="{% url 'account:site_account' %}"
```

File: user_interaction/templates/user_interaction/account_pages/site_account_page.html [hunk 13]
```diff
@@ -25,7 +25,7 @@ <h1>Account</h1>
             Real Name: <i>{{ user.first_name|default:"No real name provided" }}</i><br>
         {% endif %}
         Email: <i>{{ user.email|default:"No email provided" }}</i><br>
-        <small class="text-muted">This information can currently not be edited on your end. Please notify the <a href="mailto:uups@kotkt.nl">UUPS</a> if something needs to be changed.</small>
+        <small class="text-muted">To change your account information, click <a href="{% url 'core:user_accounts/account/edit' %}">here</a></small>
```
Comment (EricTRL) [inline user_interaction/templates/user_interaction/account_pages/site_account_page.html:28]: Same movement of urls as above
```suggestion
        <small class="text-muted">To change your account information, click <a href="{% url 'account:account_change' %}">here</a></small>
```

File: user_interaction/account_pages/config.py [hunk 14]
```diff
@@ -19,4 +19,5 @@ def get_urls(self):
             path('', SiteAccountView.as_view(config=self), name='site_account'),
             path('change-password/', AccountPasswordChangeView.as_view(config=self), name='password_change'),
             path('change-layout/', LayoutPreferencesUpdateView.as_view(config=self), name='layout_change'),
+            path('account/edit', AccountChangeView.as_view(config=self), name='account_change'),
```
Comment (EricTRL) [inline user_interaction/account_pages/config.py:22]: This causes a url of the form `/account/site/account/edit` (note the double "account"). We're also slowly transitioning to have all URLs should end with a trailing `/`.
```suggestion
            path('edit/', AccountChangeView.as_view(config=self), name='account_change'),
```

**Review by EricTRL — CHANGES_REQUESTED** [review]
A few issues regarding the moving around of URLs (note the failing test), but after this we're ready to merge!

**Comment by waiterorg** [comment]
Thank you so much for taking time to me and explaining !
I fixed urls , I was do wrong some places for not being fimiliar with the project :)
the account system has many changes from 10 days ago .
I hope everything will be fine this time !

File: core/forms.py [hunk 15]
```diff
@@ -143,3 +143,11 @@ class Meta:
...
 
     updating_user_field_name = "uploader"
+
+class AccountForm(ModelForm):
+    class Meta:
+        model = User
+        fields = ['username', 'first_name', 'email']
+        labels = {
+            'first_name': 'Real Name',
+        }
```
Comment (DutcherNL) [inline core/forms.py:153]: From a user experience perspective it is weird that we store the person name in two places (here and in membership). As long as we do though, we may need to inform the user on the way it differs. As such I propose the following addition:

```suggestion
        }
        help_texts = {
            'first_name': 'This name is not connected to the name registered on your membership. Once membership is linked, the site will use the name registered on your membership instead.',
        }
```

File: core/forms.py [hunk 16]
```diff
@@ -143,3 +143,11 @@ class Meta:
         fields = "__all__"
 
     updating_user_field_name = "uploader"
+
+class AccountForm(ModelForm):
```
Comment (DutcherNL) [inline core/forms.py:147]: Personally I'd rather see this particular form placed in `user_interaction/account_pages/forms.py` as this form will only be used in this section and thus is less relevant as part of the core. It's just a personal preference though

**Review by DutcherNL — CHANGES_REQUESTED** [review]
We really appreciate your help here and it looks really good. Sorry for making you decypher the new structure, though hopefully the massive change does make it easier now to append things anywhere in the account tabs :)
Anyway, your code looks good. Just two minor points to note, one of which is more user-communication related (because we store the users name twice). If you could change that than I'll merge your PR asap :D

**Comment by waiterorg** [comment]
Yes , the new structure seems better than before !
I changed that from .

**Review by DutcherNL — APPROVED** [review]

**Comment by DutcherNL** [comment]
The code looks good. I'll join it later this weekend.

One note I want to add for any of your current or future projects is that we advice you to write testcases. I know it is boring and if all goes well you never need it, but I've had my fair share of accidentally breaking functionality without me noticing it that test cases are really useful to implement. In this project for instance there are quite a few unique cases when dealing with dates so having test cases helps us ensure that we did not break anything when we update a piece of code somewhere in the activity_calendar module.

For more information on testing in Django read this:
https://docs.djangoproject.com/en/2.2/topics/testing/

**Comment by waiterorg** [comment]
I appreciate your advice , yes I completely with you !
I started testing and testcase a little while ago and I try to be better in test .
It makes me happy if you tell me your opinion about my test [in this demo project](https://github.com/waiterorg/WaiterShop/tree/main/WaiterShop/core/tests)

**Review by EricTRL — APPROVED** [review]
Looks good to me now, so time to merge! I also agree on @DutcherNL's comment regarding tests, but given the scale of this PR this is good enough for now!

Regarding the tests in your other project: they seem fine on a quick glance. You're also already using a library like `coverage.py` so you can check code coverage (which doesn't tell you everything about the quality of your tests, but does allow you to see which code lacks one).
I'd also recommend to integrate some Continuous Integration (CI) in that project to automatically run your tests on a push or pull request: it'll save you quite a lot of headaches if you accidentally break something! Travis is no longer free today, but GitHub Actions serves as a pretty much perfect alternative. You can look into the GitHub Actions we have for this repo in the `.github` folder for some inspiration :)

**Comment by waiterorg** [comment]
I definitely take your advice , thank you !

**Pull request merged**